azurelinuxagent/ga/cgroupapi.py

```diff
@@ -342,11 +440,85 @@ def start_extension_command(self, extension_name, command, cmd_name, timeout, sh
             with self._systemd_run_commands_lock:
                 self._systemd_run_commands.remove(process.pid)
 
-    def cleanup_legacy_cgroups(self):
+
+class SystemdCgroupApiv2(_SystemdCgroupApi):
+    """
+    Cgroup v2 interface via systemd
+    """
+    def __init__(self):
+        super(SystemdCgroupApiv2, self).__init__()
+        self._root_cgroup_path = self._get_root_cgroup_path()
+        self._controllers_enabled_at_root = self._get_controllers_enabled_at_root()
+
+    @staticmethod
+    def _get_root_cgroup_path():
         """
-        Previous versions of the daemon (2.2.31-2.2.40) wrote their PID to /sys/fs/cgroup/{cpu,memory}/WALinuxAgent/WALinuxAgent;
-        starting from version 2.2.41 we track the agent service in walinuxagent.service instead of WALinuxAgent/WALinuxAgent. If
-        we find that any of the legacy groups include the PID of the daemon then we need to disable data collection for this
-        instance (under systemd, moving PIDs across the cgroup file system can produce unpredictable results)
+        In v2, there is a unified mount point shared by all controllers. Use findmnt to get the unified mount point.
+
+          The output of findmnt is similar to
+              $ findmnt -t cgroup2 --noheadings
+              /sys/fs/cgroup cgroup2 cgroup2 rw,nosuid,nodev,noexec,relatime,nsdelegate,memory_recursiveprot
+
+        Returns None if the root cgroup cannot be determined from the output above.
+        """
+        #
+        for line in shellutil.run_command(['findmnt', '-t', 'cgroup2', '--noheadings']).splitlines():
+            # Systemd mounts the cgroup filesystem at '/sys/fs/cgroup'. The agent does not support cgroups if the
+            # filesystem is mounted elsewhere, so search specifically for '/sys/fs/cgroup' in the findmnt output.
+            match = re.search(r'(?P<path>\/sys\/fs\/cgroup)\s+cgroup2', line)
+            if match is not None:
+                root_cgroup_path = match.group('path')
+                if root_cgroup_path is not None:
+                    return root_cgroup_path
+        return None
```

Comment:
Using None as a special value makes the code harder to writer. Every piece of code that uses self._root_cgroup_path  needs to check for None. Please reconsider this design. Worst case, add a comment in init marking very clearly that self._root_cgroup_path  can be None

Reply:
There are a few places in the current cgroup implementation that return None instead of empty string when looking for paths. For example, get_process_cgroup_relative_path. All callers currently check for None. 

I'll leave it as None and add a comment in init for consistency. When I do the controller abstraction in a future PR I will revisit this.

Reply:
`When I do the controller abstraction in a future PR I will revisit this.`

Please do. Continue to use bad patterns on new code is not good, but I understand why to postpone this (as long as it gets done)